Decompile font renderer and implement "DirectRenderer"

water111 opened the pull request:
This needs cleanup and docs before it can be merged.

But we created our first frame!
![image](https://user-images.githubusercontent.com/48171810/128784833-c0370c20-658c-47fc-965e-2192722fa1b4.png)

ManDude commented on game/graphics/texture/TexturePool.cpp:
```diff
@@ -0,0 +1,195 @@
+        m_tex_converter.download_rgba8888(texture_record->data.data(), tex.dest[mip_idx],
+                                          tex.width[mip_idx], ww, hh, tex.psm, tex.clutpsm,
+                                          tex.clut_dest, size_bytes);
+        set_texture(tex.dest[mip_idx], std::move(texture_record));
```
`texture_record` is used after this line for the debug texture dumping, and this seems to be causing some sort of undefined behavior (even if debug output is disabled).

water111 replied:
Good catch this is definitely wrong.

ManDude commented on game/graphics/display.cpp:
```diff
@@ -117,7 +118,7 @@ void KillDisplay(std::shared_ptr<GfxDisplay> display) {
 
   if (GetMainDisplay() == display) {
     // killing the main display, kill all children displays too!
-    for (int i = 1; i < g_displays.size(); ++i) {
+    for (size_t i = 1; i < g_displays.size(); ++i) {
       KillDisplay(g_displays.at(i));
     }
```
This loop is actually totally wrong, my bad... This is only gonna kill every second sub-display.

water111 commented:
Todo before merging:
- Fix the issue with `link` called from windows (5 arguments)
- Figure out why textures aren't working 100% of the time on windows
- Investigate changing between fonts. This behaves strangely. For example, using `*stdcon*` and having `PAUSE` displayed at the same time usually causes `*stdcon*` to be displayed with the big font.
- dump to file and run from file?

water111 commented:
Remaining issues:
- windows textures
- clang debugger issue
- save to file

The font issue was actually a renderer bug where it was using the wrong TBP.  There were also some memory bugs related to the primitive buffer that are now fixed. 

I added synchronization on the `sync-path` to make sure the rendering is done before attempting to draw the profile bar.  This probably sacrifices a tiny bit of performance but we can get accurate renderer CPU usage now, and the game did this too, for a reason I don't understand.

water111 commented:
Performance analyzer screenshot of the code after sync-path:
![image](https://user-images.githubusercontent.com/48171810/128960629-1f182088-3026-4dbc-9b4b-a4853c835dc0.png)

In this case it's about 2% of a frame.  They cheated 4% off of the profile bar, so this seems reasonable.
